File: Assets/Bundles/Shared/Resources/TS/Projectile/Projectile.lua
-- Compiled with unity-ts v2.1.0-75
local Flamework = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Flamework
local RunUtil = require("Shared/TS/Util/RunUtil").RunUtil
local Signal = require("Shared/TS/Util/Signal").Signal
local SetTimeout = require("Shared/TS/Util/Timer").SetTimeout
local ProjectileSharedImpl = require("Shared/TS/Projectile/ProjectileSharedImpl").ProjectileSharedImpl
local Projectile
do
	Projectile = setmetatable({}, {
		__tostring = function()
			return "Projectile"
		end,
	})
	Projectile.__index = Projectile
	function Projectile.new(...)
		local self = setmetatable({}, Projectile)
		return self:constructor(...) or self
	end
	function Projectile:constructor(easyProjectile, itemType, shooter)
		self.easyProjectile = easyProjectile
		self.destroyed = false
		self.OnDestroy = Signal.new()
		self.OnCollide = Signal.new()
		self.gameObject = easyProjectile.gameObject
		self.itemType = itemType
		self.shooter = shooter
		self.OnDestroy:Connect(function()
			self.destroyed = true
		end)
		easyProjectile:onCollide(function(collision, velocity)
			local contact = collision.contacts:GetValue(0)
			local hitPoint = contact.point
			local normal = contact.normal
			local collider = contact.otherCollider
			local ignored = ProjectileSharedImpl:ShouldIgnoreCollision(self, hitPoint, normal, collider)
			if ignored then
				return nil
			end
			self.OnCollide:Fire(hitPoint, collider)
			if RunUtil:IsServer() then
				(Flamework.resolveDependency("Bundles/Server/Services/Global/Damage/Projectile/ProjectileService@ProjectileService")):HandleCollision(self, collider, hitPoint, normal, velocity)
			else
				(Flamework.resolveDependency("Bundles/Client/Controllers/Global/Damage/Projectile/ProjectileController@ProjectileController")):HandleCollision(self, collider, hitPoint, normal, velocity)
			end
		end)
		local dw = self.gameObject:GetComponent("DestroyWatcher")
		dw:OnDestroyedEvent(function()
			self.OnDestroy:Fire()
		end)
		SetTimeout(12, function()
			if not self.destroyed then
				self:Destroy()
			end
		end)
	end
	function Projectile:Destroy()
		if self.destroyed then
			return nil
		end
		self.destroyed = true
		Object:Destroy(self.gameObject)
	end
end
return {
	Projectile = Projectile,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Shared/Projectile/Projectile.ts
import { Dependency } from "@easy-games/flamework-core";
import { ProjectileController } from "Client/Controllers/Global/Damage/Projectile/ProjectileController";
import { ProjectileService } from "Server/Services/Global/Damage/Projectile/ProjectileService";
import { Entity } from "Shared/Entity/Entity";
import { ItemType } from "Shared/Item/ItemType";
import { RunUtil } from "Shared/Util/RunUtil";
import { Signal } from "Shared/Util/Signal";
import { SetTimeout } from "Shared/Util/Timer";
import { ProjectileSharedImpl } from "./ProjectileSharedImpl";

export interface ProjectileDto {
	nobId: number;
	itemType: ItemType;
	shooterEntityId?: number;
}

export class Projectile {
	public readonly gameObject: GameObject;
	public readonly itemType: ItemType;
	public readonly shooter: Entity | undefined;
	private destroyed = false;

	public readonly OnDestroy = new Signal<void>();
	/**
	 * Fired when the projectile hits something that will destroy it.
	 *
	 * For additional collide information, you should instead listen to either `ClientSignals.ProjectileHit` or `ServerSignals.ProjectileHit`
	 */
	public readonly OnCollide = new Signal<[hitPoint: Vector3, collider: Collider]>();

	constructor(private easyProjectile: EasyProjectile, itemType: ItemType, shooter: Entity | undefined) {
		this.gameObject = easyProjectile.gameObject;
		this.itemType = itemType;
		this.shooter = shooter;
		this.OnDestroy.Connect(() => {
			this.destroyed = true;
		});

		this.OnCollide.Connect((hitPoint, collider) => {
			print("[Debug]: projectile hit pos=" + tostring(hitPoint) + ", collider=" + collider.gameObject.name);
		});

		easyProjectile.onCollide((collision, velocity) => {
			const contact = collision.contacts.GetValue(0);

			let hitPoint = contact.point;
			let normal = contact.normal;
			let collider = contact.otherCollider;

			const ignored = ProjectileSharedImpl.ShouldIgnoreCollision(this, hitPoint, normal, collider);
			if (ignored) return;
			this.OnCollide.Fire(hitPoint, collider);

			if (RunUtil.IsServer()) {
				Dependency<ProjectileService>().HandleCollision(this, collider, hitPoint, normal, velocity);
			} else {
				Dependency<ProjectileController>().HandleCollision(this, collider, hitPoint, normal, velocity);
			}
		});

		const dw = this.gameObject.GetComponent<DestroyWatcher>();
		dw.OnDestroyedEvent(() => {
			this.OnDestroy.Fire();
		});

		SetTimeout(5, () => {
			if (!this.destroyed) {
				this.Destroy();
			}
		});
	}

	public Destroy(): void {
		if (this.destroyed) return;
		this.destroyed = true;

		Object.Destroy(this.gameObject);
	}
}
